pkg/auth/tokenparser_test.go

```diff
@@ -0,0 +1,111 @@
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestTokenParser(t *testing.T) {
+	// create logger and registry.
+	logger := log.New(os.Stderr, "", 0)
+	configRegistry := configuration.CreateEmptyRegistry()
+
+	// create test keys
+	tokengenerator := testutils.NewTokenManager()
+	kid0 := uuid.NewV4().String()
+	_, err := tokengenerator.AddPrivateKey(kid0)
+	require.NoError(t, err)
+	kid1 := uuid.NewV4().String()
+	_, err = tokengenerator.AddPrivateKey(kid1)
+	require.NoError(t, err)
+
+	// startup public key service
+	keysEndpointURL := tokengenerator.NewKeyServer().URL
+
+	// set the config for testing mode, the handler may use this.
+	configRegistry.GetViperInstance().Set("testingmode", true)
+	assert.True(t, configRegistry.IsTestingMode(), "testing mode not set correctly to true")
+	// set the key service url in the config
+	configRegistry.GetViperInstance().Set("auth_client.public_keys_url", keysEndpointURL)
+	assert.Equal(t, keysEndpointURL, configRegistry.GetAuthClientPublicKeysURL(), "key url not set correctly")
+
+	// create KeyManager instance.
+	keyManager, err := auth.NewKeyManager(logger, configRegistry)
+	require.NoError(t, err)
+
+	// create TokenParser instance.
+	tokenParser, err := auth.NewTokenParser(logger, configRegistry, keyManager)
+	require.NoError(t, err)
+
+	t.Run("invalid arguments to new", func(t *testing.T) {
+		_, err := auth.NewTokenParser(nil, configRegistry, keyManager)
+		require.Error(t, err)
+		_, err = auth.NewTokenParser(logger, nil, keyManager)
+		require.Error(t, err)
+		_, err = auth.NewTokenParser(logger, configRegistry, nil)
+		require.Error(t, err)
+	})
+
+	t.Run("parse valid tokens", func(t *testing.T) {
+		// create two test tokens, both valid
+		username0 := uuid.NewV4().String()
+		identity0 := &testutils.Identity{
+			ID:       uuid.NewV4(),
+			Username: username0,
+		}
+		email0 := identity0.Username + "@email.tld"
+		jwt0, err := tokengenerator.GenerateSignedToken(*identity0, kid0, testutils.WithEmailClaim(email0))
+		require.NoError(t, err)
+		username1 := uuid.NewV4().String()
+		identity1 := &testutils.Identity{
+			ID:       uuid.NewV4(),
+			Username: username1,
+		}
+		email1 := identity1.Username + "@email.tld"
+		jwt1, err := tokengenerator.GenerateSignedToken(*identity1, kid1, testutils.WithEmailClaim(email1))
+		require.NoError(t, err)
+
+		// check if the keys can be used to verify a JWT
+		var statictests = []struct {
+			name     string
+			jwt      string
+			username string
+			email    string
+		}{
+			{"valid JWT 0", jwt0, identity0.Username, email0},
+			{"valid JWT 1", jwt1, identity1.Username, email1},
+		}
+		for _, tt := range statictests {
+			t.Run(tt.name, func(t *testing.T) {
+				claims, err := tokenParser.FromString(tt.jwt)
+				require.NoError(t, err)
+				require.Equal(t, tt.username, claims.Username)
+				require.Equal(t, tt.email, claims.Email)
+			})
+		}
+	})
+
+	t.Run("parse invalid token", func(t *testing.T) {
```

Comment:
We should cover all the cases of invalid tokens:

1. No KID header in the token.
2. Token signed by unknown key.
3. Token signed by known key but the signature is invalid.
4. Signature is good but token expired.
5. Missing claims (individual tests for each missing claim)

Reply:
Fixed 96b6979. For (3): the signature implies a key, so I added a test where the signature is garbage (in addition to a test that has a valid signature from an unknown/non-kid-matching key).